General/DeviceRegistration/DeviceRegistrationWorker.swift

```diff
@@ -9,59 +9,86 @@
 import Foundation
 import Combine
 import RxSwift
-import RxCocoa
 import RealifeTech_CoreSDK
 
 class DeviceRegistrationWorker: DeviceRegistering {
 
-    var sdkReady: Bool { deviceRegisteredSubject.value }
+    var sdkReady: Bool {
+        deviceRegisteredSubject.value
+    }
     var deviceId: String { staticDeviceInformation.deviceId }
 
     private let staticDeviceInformation: StaticDeviceInformation
     private let reachabilityChecker: ReachabilityChecking
-    private let loopHandler: DeviceRegistrationLoopHandling
-    private let deviceRegisteredSubject: CurrentValueSubject<Bool, Never>
+    private let deviceProvider: DeviceProviding.Type
+    private let deviceRegistrationMaxAttempts: Int
+    private let retryRegistrationInterval: RxTimeInterval
     private let store: Storeable
+    private let subscriptionScheduler: SchedulerType
+    private let observationScheduler: SchedulerType
+    private let deviceRegisteredSubject = CurrentValueSubject<Bool, Never>(false)
     private let storeKey = "DeviceRegistrationState"
+    private let bag = DisposeBag()
 
     var device: Device {
-        Device(deviceId: staticDeviceInformation.deviceId,
-               model: staticDeviceInformation.deviceModel,
-               sdkVersion: "SDK_" + staticDeviceInformation.sdkVersion,
-               osVersion: staticDeviceInformation.osVersion,
-               bluetoothOn: reachabilityChecker.isBluetoothConnected,
-               wifiConnected: reachabilityChecker.isConnectedToWifi)
+        Device(
+            deviceId: staticDeviceInformation.deviceId,
+            model: staticDeviceInformation.deviceModel,
+            sdkVersion: "SDK_" + staticDeviceInformation.sdkVersion,
+            osVersion: staticDeviceInformation.osVersion,
+            bluetoothOn: reachabilityChecker.isBluetoothConnected,
+            wifiConnected: reachabilityChecker.isConnectedToWifi)
     }
 
     init(
         staticDeviceInformation: StaticDeviceInformation,
         reachabilityChecker: ReachabilityChecking,
-        loopHandler: DeviceRegistrationLoopHandling,
-        deviceRegisteredSubject: CurrentValueSubject<Bool, Never>,
-        store: Storeable
+        deviceProvider: DeviceProviding.Type = DeviceRepository.self,
+        deviceRegistrationMaxAttempts: Int = .max,
```

Comment:
could using .max mean that it'll try to register this many times: [9223372036854775807](https://www.tutorialspoint.com/swift/swift_data_types.htm)?

Reply:
Yes, using .max keeps retrying until the user quits the app. I think it's risky to set the default value as max but couldn't find any suggested retry time on the mobile SDK doc. Assuming the retry interval is 10 seconds by default, setting the maximum retry attempts to 60 times takes 10 minutes to see the final error.

How do you feel about we change the default value to 60 or 360 attempts? It takes 10 minutes or 1 hour to retry the last time. It makes more sense than .max in my opinion as no users will wait that long